File: packages/python-sdk/src/opensend/segments.py
"""Segments resource for the OpenSend Python SDK."""

from __future__ import annotations

from typing import Optional, cast

from ._http import HttpClient
from ._types import (
    CreateSegmentPayload,
    DeleteSegmentResponse,
    ListOptions,
    SegmentContactListResponse,
    SegmentListOptions,
    SegmentListResponse,
    SegmentResponse,
)


class SegmentsResource:
    """CRUD + contacts-within-segment for the /segments namespace."""

    def __init__(self, client: HttpClient) -> None:
        self._client = client

    def create(self, payload: CreateSegmentPayload) -> SegmentResponse:
        """Create a new contact segment."""
        return cast(SegmentResponse, self._client.request("POST", "/segments", payload))

    def list(self, options: Optional[SegmentListOptions] = None) -> SegmentListResponse:
        """List segments with optional pagination and search."""
        opts = options or {}
        query: dict[str, str] = {}
        if opts.get("limit") is not None:
            query["limit"] = str(opts["limit"])
        if opts.get("after"):
            query["after"] = opts["after"]  # type: ignore[assignment]
        if opts.get("search"):
            query["search"] = opts["search"]  # type: ignore[assignment]
        return cast(
            SegmentListResponse,
            self._client.request("GET", "/segments", params=query or None),
        )

    def get(self, segment_id: str) -> SegmentResponse:
        """Retrieve a segment by ID."""
        return cast(SegmentResponse, self._client.request("GET", f"/segments/{segment_id}"))

    def delete(self, segment_id: str) -> DeleteSegmentResponse:
        """Delete a segment by ID."""
        return cast(
            DeleteSegmentResponse,
            self._client.request("DELETE", f"/segments/{segment_id}"),
        )

    def list_contacts(
        self, segment_id: str, options: Optional[ListOptions] = None
    ) -> SegmentContactListResponse:
        """List contacts that belong to a given segment."""
        opts = options or {}
        query: dict[str, str] = {}
        if opts.get("limit") is not None:
            query["limit"] = str(opts["limit"])
        if opts.get("after"):
            query["after"] = opts["after"]  # type: ignore[assignment]
        return cast(
            SegmentContactListResponse,
            self._client.request(
                "GET", f"/segments/{segment_id}/contacts", params=query or None
            ),
        )

File: packages/ruby-sdk/lib/opensend/resources/segments.rb
# frozen_string_literal: true

module OpenSend
  module Resources
    class Segments
      def initialize(http)
        @http = http
      end

      # POST /segments
      def create(params)
        @http.post("/segments", stringify(params))
      end

      # GET /segments
      def list(limit: nil, after: nil, search: nil)
        params = {}
        params[:limit] = limit unless limit.nil?
        params[:after] = after if after
        params[:search] = search if search
        @http.get("/segments", params)
      end

      # GET /segments/:id
      def get(id)
        @http.get("/segments/#{id}")
      end

      # DELETE /segments/:id
      def delete(id)
        @http.delete("/segments/#{id}")
      end

      # GET /segments/:id/contacts
      def list_contacts(id, limit: nil, after: nil)
        params = {}
        params[:limit] = limit unless limit.nil?
        params[:after] = after if after
        @http.get("/segments/#{id}/contacts", params)
      end

      private

      def stringify(params)
        return params unless params.is_a?(Hash)

        params.transform_keys(&:to_s)
      end
    end
  end
end
